include/mem_native.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ * Copyright (C) 2020-2020  The dosbox-staging team
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+/*  These functions read and write 16, 32, and 64 bit uint's to
+ *  and from 8-bit memory positions using defined alignment . Use
+ *  these functions instead of C-style pointer type casts to larger
+ *  types, which are not alignment-safe.
+ */
+
+#ifndef DOSBOX_MEM_NATIVE_H
+#define DOSBOX_MEM_NATIVE_H
+
+#include <cstdint>
+#include <cstring>
+
+/*  Read a uint16 from 8-bit native byte-ordered memory. Use this
+ *  instead of *(uint16_t*)(8_bit_ptr) or *(uint16_t*)(8_bit_ptr + offset)
+ */
+static INLINE uint16_t read_uint16(const uint8_t *arr)
```

Comment:
I am still not entirely sure about naming these functions `read_*` / `read_*_at`, but haven't came up with better names for it, so let's keep them in for now (especially since they are not used yet, so we can easily change them).

After a bit of consideration, I think we should stop using `INLINE` here - I don't think it helps compiler with inlining these functions at all. Using INLINE would also require including `config.h`.

Reason for dropping `INLINE`: when using GCC our config.h translates it to `inline __attribute__((always_inline))`; there are conflicting descriptions of attribute always_inline, but GNU documentation is clear: [link](https://gcc.gnu.org/onlinedocs/gcc/Inline.html) - this attribute exists to optimize code even when building without optimizations… so let's drop it - it will only make our perf analysis less accurate. I suggest using `static inline` for functions in this header.

Reply:
Absolutely; the more we can defer to the compiler, the better. I've switched over to `inline` now.

Compared to ~2003, today the latest GCC and Clang releases have their optimization decisions coming from huge datasets and runs by the big players like Google and Facebook; where every cycle counts.

Yup, can always adjust the names ahead of the next PR.